File: sdks/python/src/verdocs/resources/members.py
"""Organization member operations (js-sdk: Organizations/Members.ts).

A member is a profile with organization access. All calls here operate on
the caller's current organization; the admin-only calls (everything except
list) require an admin or owner role.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..models.core import Profile
from ..models.organizations import MemberCreateParams, MemberCreateResponse, MemberUpdateParams

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_MEMBERS_PATH = "/v2/organization-members"


def _write_body(params: MemberCreateParams | MemberUpdateParams) -> dict[str, Any]:
    return params.model_dump(mode="json", exclude_unset=True)


class Members:
    """Organization member calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def list(self) -> list[Profile]:
        """Get the members of the caller's organization via GET /v2/organization-members.

        Contacts and guest profiles are excluded. When the caller is an admin
        or owner, each profile carries its joined user record (for lock-state
        management); other callers get plain profiles. Mirrors js-sdk
        getOrganizationMembers.

        Returns:
            The organization's members, sorted by last then first name.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.getOrganizationMembers
        @sdkGroup Member
        @sdkPage Endpoints
        """
        response = self._endpoint._request("GET", _MEMBERS_PATH)
        return [Profile.model_validate(entry) for entry in response.json()]

    def create(self, params: MemberCreateParams) -> MemberCreateResponse:
        """Create a member directly, bypassing the invite flow, via POST /v2/organization-members.

        If no user account exists for the email, one is created; omit
        password to have the server generate one and return it. Mirrors
        js-sdk createOrganizationMember, which types the response as a bare
        profile; the wire wraps it (see MemberCreateResponse).

        Example:
            result = endpoint.members.create(
                MemberCreateParams(email="a@b.com", first_name="A", last_name="B", roles=["member"])
            )
            initial_password = result.password

        Args:
            params: Details for the new member.

        Returns:
            The new profile, a user summary, and the generated password when
            the server created the account.

        Raises:
            VerdocsAPIError: The API rejected the request (400 when the user
                is already a member).
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.createOrganizationMember
        @sdkGroup Member
        @sdkPage Endpoints
        """
        response = self._endpoint._request("POST", _MEMBERS_PATH, json=_write_body(params))
        return MemberCreateResponse.model_validate(response.json())

    def update(self, profile_id: str, params: MemberUpdateParams) -> Profile:
        """Update a member's roles via PATCH /v2/organization-members/{profile_id}.

        Callers may not update themselves, and only an owner may grant the
        owner role. Mirrors js-sdk updateOrganizationMember; note the
        deployed schema accepts roles only (see MemberUpdateParams).

        Args:
            profile_id: The profile to update.
            params: The fields to change.

        Returns:
            The updated profile.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.updateOrganizationMember
        @sdkGroup Member
        @sdkPage Endpoints
        """
        response = self._endpoint._request("PATCH", f"{_MEMBERS_PATH}/{profile_id}", json=_write_body(params))
        return Profile.model_validate(response.json())

    def delete(self, profile_id: str) -> None:
        """Remove a member via DELETE /v2/organization-members/{profile_id}.

        The member's envelopes, templates, and recipient records are
        reassigned to the caller; their signatures, API keys, and similar
        personal records are deleted. Callers may not delete themselves.
        Mirrors js-sdk deleteOrganizationMember.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            profile_id: The profile to remove.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.deleteOrganizationMember
        @sdkGroup Member
        @sdkPage Endpoints
        """
        self._endpoint._request("DELETE", f"{_MEMBERS_PATH}/{profile_id}")

    def lock(self, profile_id: str, reason: str) -> Profile:
        """Lock a member's user account via PUT /v2/organization-members/{profile_id}.

        A locked member cannot sign in until unlocked or until they complete
        a password reset. Callers may not lock themselves, the target must
        have a linked user account, and only an owner may lock another owner.
        Mirrors js-sdk lockOrganizationMember.

        Args:
            profile_id: The profile to lock.
            reason: Why the account is being locked (1-255 chars); stored on
                the user record and shown to admins.

        Returns:
            The member's profile with the joined user record.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.lockOrganizationMember
        @sdkGroup Member
        @sdkPage Endpoints
        """
        response = self._endpoint._request(
            "PUT", f"{_MEMBERS_PATH}/{profile_id}", json={"action": "lock", "reason": reason}
        )
        return Profile.model_validate(response.json())

    def unlock(self, profile_id: str) -> Profile:
        """Unlock a member's user account via PUT /v2/organization-members/{profile_id}.

        Clears the locked flag, the lock reason, and the failed-login
        counter. Callers may not unlock themselves and the target must have a
        linked user account. Mirrors js-sdk unlockOrganizationMember.

        Args:
            profile_id: The profile to unlock.

        Returns:
            The member's profile with the joined user record.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.

        @sdkOperation member.unlockOrganizationMember
        @sdkGroup Member
        @sdkPage Endpoints
        """
        response = self._endpoint._request("PUT", f"{_MEMBERS_PATH}/{profile_id}", json={"action": "unlock"})
        return Profile.model_validate(response.json())


class AsyncMembers:
    """Organization member calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def list(self) -> list[Profile]:
        """Get the members of the caller's organization via GET /v2/organization-members.

        Contacts and guest profiles are excluded. When the caller is an admin
        or owner, each profile carries its joined user record (for lock-state
        management); other callers get plain profiles. Mirrors js-sdk
        getOrganizationMembers.

        Returns:
            The organization's members, sorted by last then first name.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _MEMBERS_PATH)
        return [Profile.model_validate(entry) for entry in response.json()]

    async def create(self, params: MemberCreateParams) -> MemberCreateResponse:
        """Create a member directly, bypassing the invite flow, via POST /v2/organization-members.

        If no user account exists for the email, one is created; omit
        password to have the server generate one and return it. Mirrors
        js-sdk createOrganizationMember, which types the response as a bare
        profile; the wire wraps it (see MemberCreateResponse).

        Example:
            result = await endpoint.members.create(
                MemberCreateParams(email="a@b.com", first_name="A", last_name="B", roles=["member"])
            )
            initial_password = result.password

        Args:
            params: Details for the new member.

        Returns:
            The new profile, a user summary, and the generated password when
            the server created the account.

        Raises:
            VerdocsAPIError: The API rejected the request (400 when the user
                is already a member).
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", _MEMBERS_PATH, json=_write_body(params))
        return MemberCreateResponse.model_validate(response.json())

    async def update(self, profile_id: str, params: MemberUpdateParams) -> Profile:
        """Update a member's roles via PATCH /v2/organization-members/{profile_id}.

        Callers may not update themselves, and only an owner may grant the
        owner role. Mirrors js-sdk updateOrganizationMember; note the
        deployed schema accepts roles only (see MemberUpdateParams).

        Args:
            profile_id: The profile to update.
            params: The fields to change.

        Returns:
            The updated profile.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", f"{_MEMBERS_PATH}/{profile_id}", json=_write_body(params))
        return Profile.model_validate(response.json())

    async def delete(self, profile_id: str) -> None:
        """Remove a member via DELETE /v2/organization-members/{profile_id}.

        The member's envelopes, templates, and recipient records are
        reassigned to the caller; their signatures, API keys, and similar
        personal records are deleted. Callers may not delete themselves.
        Mirrors js-sdk deleteOrganizationMember.

        The API answers with a status marker that nothing consumes, so this
        returns None and relies on exceptions for failure.

        Args:
            profile_id: The profile to remove.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        await self._endpoint._request("DELETE", f"{_MEMBERS_PATH}/{profile_id}")

    async def lock(self, profile_id: str, reason: str) -> Profile:
        """Lock a member's user account via PUT /v2/organization-members/{profile_id}.

        A locked member cannot sign in until unlocked or until they complete
        a password reset. Callers may not lock themselves, the target must
        have a linked user account, and only an owner may lock another owner.
        Mirrors js-sdk lockOrganizationMember.

        Args:
            profile_id: The profile to lock.
            reason: Why the account is being locked (1-255 chars); stored on
                the user record and shown to admins.

        Returns:
            The member's profile with the joined user record.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request(
            "PUT", f"{_MEMBERS_PATH}/{profile_id}", json={"action": "lock", "reason": reason}
        )
        return Profile.model_validate(response.json())

    async def unlock(self, profile_id: str) -> Profile:
        """Unlock a member's user account via PUT /v2/organization-members/{profile_id}.

        Clears the locked flag, the lock reason, and the failed-login
        counter. Callers may not unlock themselves and the target must have a
        linked user account. Mirrors js-sdk unlockOrganizationMember.

        Args:
            profile_id: The profile to unlock.

        Returns:
            The member's profile with the joined user record.

        Raises:
            NotFoundError: No such member in the caller's organization.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PUT", f"{_MEMBERS_PATH}/{profile_id}", json={"action": "unlock"})
        return Profile.model_validate(response.json())

File: sdks/csharp/src/Verdocs.Sdk/Resources/Members.cs
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Organization member calls, reached through the endpoint's Members property. A member is a
/// profile with access to the caller's organization; contacts (profiles without access) live
/// under the Contacts resource instead.
/// </summary>
public sealed class Members
{
    private readonly VerdocsEndpoint _endpoint;

    internal Members(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the members of the caller's organization, sorted by name. Admins and owners also
    /// receive each member's joined user record (for lock-state management); other callers
    /// get the plain profiles.
    ///
    /// <example>
    /// <code>
    /// var members = await endpoint.Members.ListAsync();
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The organization's member profiles.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    /// <sdkOperation>member.getOrganizationMembers</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public async Task<IReadOnlyList<Profile>> ListAsync(CancellationToken cancellationToken = default)
    {
        return await _endpoint.SendAsync<List<Profile>>(HttpMethod.Get, "/v2/organization-members", null, cancellationToken)
            .ConfigureAwait(false);
    }

    /// <summary>
    /// Provisions a member directly, bypassing the invitation flow. The caller must be an
    /// admin. When no user exists for the email, one is created with the supplied password,
    /// or with a generated password returned in the response.
    ///
    /// <example>
    /// <code>
    /// var created = await endpoint.Members.CreateAsync(new CreateMemberRequest
    /// {
    ///     Email = "new.member@example.com",
    ///     FirstName = "New",
    ///     LastName = "Member",
    ///     Roles = ["member"],
    /// });
    /// var initialPassword = created.Password;
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">Details for the new member.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The new profile, the backing user account, and a generated password when one was created.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the user is already a member.</exception>
    /// <sdkOperation>member.createOrganizationMember</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<CreateMemberResponse> CreateAsync(CreateMemberRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<CreateMemberResponse>(HttpMethod.Post, "/v2/organization-members", request, cancellationToken);
    }

    /// <summary>
    /// Updates a member's roles. The caller must be an admin (an owner, to grant the owner
    /// role) and may not update their own profile here.
    ///
    /// <example>
    /// <code>
    /// var updated = await endpoint.Members.UpdateAsync(
    ///     "d2338742-f3a1-465b-8592-806587413cc1",
    ///     new UpdateMemberRequest { Roles = ["admin"] });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="profileId">The member's profile ID.</param>
    /// <param name="request">The changes to apply.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated profile.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the caller targeted their own profile.</exception>
    /// <sdkOperation>member.updateOrganizationMember</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Profile> UpdateAsync(string profileId, UpdateMemberRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Profile>(
            HttpMethod.Patch, "/v2/organization-members/" + Uri.EscapeDataString(profileId), request, cancellationToken);
    }

    /// <summary>
    /// Removes a member from the caller's organization. The caller must be an admin. The
    /// member's envelopes, recipient records, and templates are reassigned to the caller;
    /// their signatures, API keys, and group memberships are deleted.
    ///
    /// <example>
    /// <code>
    /// await endpoint.Members.DeleteAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="profileId">The member's profile ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>A task that completes when the member is removed.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the profile was not found.</exception>
    /// <sdkOperation>member.deleteOrganizationMember</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task DeleteAsync(string profileId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        return _endpoint.SendVoidAsync(
            HttpMethod.Delete, "/v2/organization-members/" + Uri.EscapeDataString(profileId), null, cancellationToken);
    }

    /// <summary>
    /// Locks a member's account so they cannot sign in until an admin unlocks them or they
    /// complete a password reset. The caller must be an admin (an owner, to lock another
    /// owner), may not lock themselves, and the member must have a linked user account.
    ///
    /// <example>
    /// <code>
    /// var locked = await endpoint.Members.LockAsync("d2338742-f3a1-465b-8592-806587413cc1", "Suspicious activity");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="profileId">The member's profile ID.</param>
    /// <param name="reason">Why the account is being locked (1 to 255 characters). Stored on the user record and shown to admins.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The member's profile with the joined user record.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the member has no user account.</exception>
    /// <sdkOperation>member.lockOrganizationMember</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Profile> LockAsync(string profileId, string reason, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        ArgumentException.ThrowIfNullOrEmpty(reason);
        return _endpoint.SendAsync<Profile>(
            HttpMethod.Put,
            "/v2/organization-members/" + Uri.EscapeDataString(profileId),
            new { action = "lock", reason },
            cancellationToken);
    }

    /// <summary>
    /// Unlocks a member whose account was locked by an admin or by too many failed sign-in
    /// attempts. Clears the lock reason and the failure counter. The caller must be an admin
    /// and may not target themselves.
    ///
    /// <example>
    /// <code>
    /// var unlocked = await endpoint.Members.UnlockAsync("d2338742-f3a1-465b-8592-806587413cc1");
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="profileId">The member's profile ID.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The member's profile with the joined user record.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the member has no user account.</exception>
    /// <sdkOperation>member.unlockOrganizationMember</sdkOperation>
    /// <sdkGroup>Member</sdkGroup>
    /// <sdkPage>Endpoints</sdkPage>
    public Task<Profile> UnlockAsync(string profileId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        return _endpoint.SendAsync<Profile>(
            HttpMethod.Put,
            "/v2/organization-members/" + Uri.EscapeDataString(profileId),
            new { action = "unlock" },
            cancellationToken);
    }
}
